File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans(n,1);
        int temp = nums[0];
        for(int i = 1 ; i < n ; i++){
            ans[i]*=temp;
            temp*=nums[i];
        }
        temp = nums[n-1];
        for(int i = n-2 ; i >=0  ; i--){
            ans[i]*=temp;
            temp*=nums[i];
        }
        return ans;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.go
func productExceptSelf(nums []int) []int {
    n := len(nums)
    pref := make([]int, n)
    suff := make([]int, n)
    ans := make([]int, n)

    mult := 1
    for i := 0; i < n; i++ {
        mult *= nums[i]
        pref[i] = mult
    }

    mult = 1
    for i := n - 1; i >= 0; i-- {
        mult *= nums[i]
        suff[i] = mult
    }

    for i := 0; i < n; i++ {
        if i == 0 {
            ans[i] = suff[1]
        } else if i == n-1 {
            ans[i] = pref[n-2]
        } else {
            ans[i] = pref[i-1] * suff[i+1]
        }
    }

    return ans
}
